src/contract.rs

```diff
@@ -17,37 +18,67 @@ struct Staking {
     all_produced: u128,
     reward_produced: u128,
     stakers: HashMap<ActorId, Staker>,
+
+    transactions: BTreeMap<ActorId, Transaction<StakingAction>>,
+    current_tid: TransactionId,
 }
 
 static mut STAKING: Option<Staking> = None;
 const DECIMALS_FACTOR: u128 = 10_u128.pow(20);
 
-/// Transfers `amount` tokens from `sender` account to `recipient` account.
-/// Arguments:
-/// * `from`: sender account
-/// * `to`: recipient account
-/// * `amount`: amount of tokens
-async fn transfer_tokens(
-    token_address: &ActorId,
-    from: &ActorId,
-    to: &ActorId,
-    amount_tokens: u128,
-) {
-    msg::send_for_reply(
-        *token_address,
-        FTAction::Transfer {
-            from: *from,
-            to: *to,
+impl Staking {
+    /// Transfers `amount` tokens from `sender` account to `recipient` account.
+    /// Arguments:
+    /// * `from`: sender account
+    /// * `to`: recipient account
+    /// * `amount`: amount of tokens
+    async fn transfer_tokens(
+        &mut self,
+        token_address: &ActorId,
+        from: &ActorId,
+        to: &ActorId,
+        amount_tokens: u128,
+    ) -> Result<Vec<u8>, Error> {
+        let payload = Action::Transfer {
+            sender: *from,
+            recipient: *to,
             amount: amount_tokens,
-        },
-        0,
-    )
-    .expect("Error in sending message")
-    .await
-    .expect("Error in transfer");
-}
+        }
+        .encode();
+
+        let transaction_id = self.current_tid;
+        self.current_tid = self.current_tid.saturating_add(99);
+
+        let payload = FTokenAction::Message {
+            transaction_id,
+            payload,
+        };
+        gstd::debug!(
+            "transfer_tokens() payload = {:?}, tid = {}, token_address = {:?}",
+            payload,
+            transaction_id,
+            token_address
+        );
+
+        let result = msg::send_for_reply(*token_address, payload, 0)
+            .map_err(|e| {
+                gstd::debug!("Error at transfer = {:?}", e);
+                Error::TransferTokens
+            })?
+            .await
+            .map_err(|e| {
+                gstd::debug!("Error at await = {:?}", e);
+                Error::TransferTokens
+            });
```

Comment:
I think an implementation of `From<ContractError>` for contract's `Error` can be provided, and then these `map_err`s can be removed.
```suggestion
            .?.await;
```